### PR TITLE
refactor: Rejection flow 

### PR DESCRIPTION
**Pull request by HenrikSommer**
## Description

Following the refactoring of command handlers and as part of [#1306](https://app.zenhub.com/workspaces/team-volt-60a61059028fd7000e49956e/issues/energinet-datahub/geh-charges/1306) we secure ordering of operations in the asynchronous rejection flow.

In addition, this pull request address an issue in which some merged errormessages contained the `unknown` tag, and tests are added to secure this issue will not occur again when more validation rules are added.

## References

* #1306

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/DocumentValidation/Factories/ChargeCommandDocumentValidationRulesFactory.cs [hunk 1]
```diff
@@ -42,20 +42,20 @@ public async Task<IValidationRuleSet> CreateRulesAsync(ChargeCommand chargeComma
...
-            var rules = new List<IValidationRule>
+            var rules = new List<IValidationRuleContainer>
             {
-                new CommandSenderMustBeAnExistingMarketParticipantRule(sender),
-                new BusinessReasonCodeMustBeUpdateChargeInformationOrChargePricesRule(documentDto),
-                new DocumentTypeMustBeRequestChangeOfPriceListRule(documentDto),
-                new RecipientIsMandatoryTypeValidationRule(documentDto),
-                new SenderIsMandatoryTypeValidationRule(documentDto),
-                new RecipientMustBeDdzRule(documentDto),
+                new CommandValidationRuleContainer(new CommandSenderMustBeAnExistingMarketParticipantRule(sender)),
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/DocumentValidation/Factories/ChargeCommandDocumentValidationRulesFactory.cs:53]: Hvae you considered calling it DocumentValidationRuleContainer?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationErrorFactory.cs [hunk 2]
```diff
@@ -18,25 +18,28 @@ namespace GreenEnergyHub.Charges.Domain.Dtos.Validation
...
             return rule =>
             {
-                if (rule is IValidationRuleForOperation validationRuleForOperation)
+                if (rule is IOperationValidationRuleContainer operationValidationRuleContainer &&
+                    rule.ValidationRule is IValidationRuleWithExtendedData validationRuleWithExtendedData)
                 {
                     return new ValidationError(
-                        rule.ValidationRuleIdentifier,
-                        null);
+                        rule.ValidationRule.ValidationRuleIdentifier,
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationErrorFactory.cs:29]: Can we add `ValidationErrorFactoryTests`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/DocumentValidation/Factories/ChargeCommandDocumentValidationRulesFactory.cs (hunk 1 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeCommands/Validation/DocumentValidation/Factories/ChargeCommandDocumentValidationRulesFactory.cs:53]: I didn't, but it is a good idea :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ChargeOperationInputValidationRulesFactoryTests.cs [hunk 3]
```diff
@@ -32,6 +33,25 @@ namespace GreenEnergyHub.Charges.Tests.Domain.Dtos.ChargeCommands.Validation.Inp
     [UnitTest]
     public class ChargeOperationInputValidationRulesFactoryTests
     {
+        [Fact]
+        public void CreateRulesForChargeCommand_ShouldContainRulesTest()
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ChargeOperationInputValidationRulesFactoryTests.cs:37]:
```suggestion
        public void CreateRulesForChargeOperation_ShouldContainRulesTest()
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationErrorFactory.cs (hunk 2 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/Validation/ValidationErrorFactory.cs:29]: I'll give it a go.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ChargeOperationInputValidationRulesFactoryTests.cs (hunk 3 above)
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ChargeOperationInputValidationRulesFactoryTests.cs:37]: The test below seems identical with this one?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeMustExistRuleTests.cs [hunk 4]
```diff
@@ -28,26 +26,17 @@ public class ChargeMustExistRuleTests
     {
         [Theory]
         [InlineAutoMoqData]
-        public void IsValid_WhenCalledWithMeteringPoint_ReturnsTrue(Charge charge, ChargeLinkDto chargeLinkDto)
+        public void IsValid_WhenCalledWithMeteringPoint_ReturnsTrue(Charge charge)
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeMustExistRuleTests.cs:29]:
```suggestion
        public void IsValid_WhenCalledWithCharge_ReturnsTrue(Charge charge)
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeMustExistRuleTests.cs [hunk 5]
```diff
@@ -28,26 +26,17 @@ public class ChargeMustExistRuleTests
...
-            var sut = new ChargeMustExistRule(charge, chargeLinkDto);
+            var sut = new ChargeMustExistRule(charge);
             sut.IsValid.Should().BeTrue();
         }
 
-        [Theory]
-        [InlineAutoMoqData]
-        public void IsValid_WhenCalledWithNullMeteringPoint_ReturnsFalse(ChargeLinkDto chargeLinkDto)
+        [Fact]
+        public void IsValid_WhenCalledWithNullMeteringPoint_ReturnsFalse()
```
Comment (prtandrup) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeMustExistRuleTests.cs:36]:
```suggestion
        public void IsValid_WhenCalledWithNullCharge_ReturnsFalse()
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ChargeOperationInputValidationRulesFactoryTests.cs (hunk 3 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeCommands/Validation/InputValidation/ChargeOperationInputValidationRulesFactoryTests.cs:37]: True, and I have removed one of them

**Review by prtandrup — APPROVED** [review]
Good job 👍

**Pull request merged**